modules/gapi/test/util/variant_tests.cpp

```diff
@@ -486,4 +500,209 @@ TEST(Variant, EXT_IndexOf)
+
+    var = float{6.0};
+    EXPECT_TRUE(cv::util::visit(visitor, var));
+    EXPECT_EQ(ss.str(), std::string("0:42,1:1,2:a,3:6,"));
+
+    var = test_validation::MyType{};
+    EXPECT_TRUE(cv::util::visit(visitor, var));
+    EXPECT_EQ(ss.str(), std::string("0:42,1:1,2:a,3:6,4:MyType,"));
+
+    var = test_validation::MyClass{};
+    EXPECT_TRUE(cv::util::visit(visitor, var));
+    EXPECT_EQ(ss.str(), std::string("0:42,1:1,2:a,3:6,4:MyType,5:MyClass,"));
+}
+
+TEST(Variant, StaticVisitor)
+{
+    using V = cv::util::variant<int, double, char, float, test_validation::MyType, test_validation::MyClass>;
+    V var{42};
+    std::stringstream ss;
+    test_validation::MyNoParamStaticVisitor visitor(ss);
+
+    cv::util::visit(visitor, var);
+    EXPECT_EQ(ss.str(), std::string("42,"));
+
+    var = double{1.0};
+    cv::util::visit(visitor, var);
+    EXPECT_EQ(ss.str(), std::string("42,1,"));
+
+    var = char{'a'};
+    cv::util::visit(visitor, var);
+    EXPECT_EQ(ss.str(), std::string("42,1,a,"));
+
+    var = float{6.0};
+    cv::util::visit(visitor, var);
+    EXPECT_EQ(ss.str(), std::string("42,1,a,6,"));
+
+    var = test_validation::MyType{};
+    cv::util::visit(visitor, var);
+    EXPECT_EQ(ss.str(), std::string("42,1,a,6,MyType,"));
+
+    var = test_validation::MyClass{};
+    cv::util::visit(visitor, var);
+    EXPECT_EQ(ss.str(), std::string("42,1,a,6,MyType,MyClass,"));
+}
+
+TEST(Variant, StaticIndexedVisitor)
+{
+    using V = cv::util::variant<int, double, char, float, test_validation::MyType, test_validation::MyClass>;
+    V var{42};
+
+    std::stringstream ss;
+    cv::util::visit(test_validation::MyNoParamStaticIndexedVisitor {ss}, var);
+    EXPECT_EQ(ss.str(), std::string("0:42,"));
+
+    var = double{1.0};
+    cv::util::visit(test_validation::MyNoParamStaticIndexedVisitor (ss), var);
+    EXPECT_EQ(ss.str(), std::string("0:42,1:1,"));
+
+    var = char{'a'};
+    cv::util::visit(test_validation::MyNoParamStaticIndexedVisitor (ss), var);
+    EXPECT_EQ(ss.str(), std::string("0:42,1:1,2:a,"));
+
+    var = float{6.0};
+    cv::util::visit(test_validation::MyNoParamStaticIndexedVisitor (ss), var);
+    EXPECT_EQ(ss.str(), std::string("0:42,1:1,2:a,3:6,"));
+
+    var = test_validation::MyType{};
+    cv::util::visit(test_validation::MyNoParamStaticIndexedVisitor (ss), var);
+    EXPECT_EQ(ss.str(), std::string("0:42,1:1,2:a,3:6,4:MyType,"));
+
+    var = test_validation::MyClass{};
+    cv::util::visit(test_validation::MyNoParamStaticIndexedVisitor (ss), var);
+    EXPECT_EQ(ss.str(), std::string("0:42,1:1,2:a,3:6,4:MyType,5:MyClass,"));
+}
+
+
+TEST(Variant, LambdaVisitor)
+{
+    using V = cv::util::variant<int, double, char, float, test_validation::MyType, test_validation::MyClass>;
+    V var{42};
+    {
+        cv::util::visit(cv::overload_lambdas(
+                [](int value) {
+                    EXPECT_EQ(value, 42);
+                },
+                [](double) {
+                    EXPECT_TRUE(false);
```

Comment:
I believe `ADD_FAILURE()` or `FAIL` is the right thing in this case.

Also, should this list cover all `Types..` range? What if we want to handle only this or that case? (e.g. `T1` and `T3` only)? Can a "default" lambda (with no value passed in, or with variant itself passed in) be introduced here?

Reply:
NO, it's limitation for C++11 lambda. In C++14 generic lambda can helps to reduce amount of code
That's because I insist to keep ClassVariant implementation too

>> Can a "default" lambda

It is possible to implement `Bypass` struct here

```
namespace opencv {
namespace utils {
template<class T>
class Bypass {
operator(T val) {}    <---nothing to do
};
}
}

temlplate<class T...> overload_lambdas_set<Bypass<T>...> make_bypass() {...}
```

Then
```
cv::util::visit(cv::overload_lambdas([] (int) {}, make_bypass<double,... ???,...>(), ...);
```

But I'm not sure about usefulness of this case. maybe as theoretic exploration pros-cons

Reply:
My comment was not about the template lambdas but about _not_ specifying some overloads

Reply:
It won't compile, i believe

Also i believe that it is not extendable, error proneous and dangerous:   Because after planned extension of variant types list ( adding new type for example) you will not receive any error and newly added type would not be processed silently